README.md
# readers_writer_lock

A clone of the `SHLock` class (and associated helper classes) from the last published version
(version 0.3.1) of Python `threading2` from PyPi, written by __Ryan Kelly__ (see
[the original project](https://github.com/rfk/threading2)).  I created this clone
since we needed to continue using the `SHLock` class in Python 3.x, and the original had stopped
being maintained (it is archived), and was stuck at Python 2.x.

This project keeps only the `SHLock` class (and its essential helper classes and components)
from the original `threading2` package.  The only modifications I have made is to port this code
to Python 3.8.

Caveats
=======
- Note that `SHLock.acquire()` does _not_ work like other lock classes in `threading`
  when the `blocking` parameter is set to `False` (and the `shared` parameter is `True`);
  in fact, it is rather useless to set the `blocking` parameter to `False`, since it is not
  possible to determine whether the lock was acquired or not in that case.

Support Considerations
======================
Currently, I have no plans to support this long-term, since this is a temporary measure to
port our own code to Python 3.x as painlessly as possible; in future, we might move to using
a different readers-writers lock implementation in Python, at which time I will drop support
for this.



ChangeLog.txt

v0.5.1:

    * Fix typo in MANIFEST; bump patch version number.

v0.5.0:

    * Port rwlock to Python 3.8.x.  No longer works with Python 2.x.

v0.4.0:

    * New clone of original project; keeps only the SHLock class and
      its dependencies.

v0.3.1:

    * posix: don't try to use sched_setaffinity for setting thread affinity.
      It requires the os-level thread id which we do not know without hackery.

v0.3.0:

    * ThreadGroup: use weakrefs internally to avoid creating memory leaks.
    * SHLock: fix cleanup logic when acquire() times out

v0.2.1:

    * rework SHLock class to avoid deadlocks, be fair and re-entrant.
    * ensure that the thread addinity is set properly on thread startup.

v0.2.0:

    * add SHLock class for shared/exclusive (also known as read/write) locks.

v0.1.4:

    * fix silly typo when loading libc (it's a miracle previous versions
      worked at all!)

v0.1.3:

    * better CPUSet <-> bitmask conversion
    * posix: dynamically calculate size of _cpuset struct.

v0.1.1:

    * support for setting process- and thread-level CPU affinity.

v0.1.0:

    * initial release - you might say *everything* has changed...



test/run_unit_tests.py
# Copyright (C) 2020 Ankan Pramanick - All rights reserved.
"""
Runs the unit tests.

Note that this MUST be invoked from the directory in which this script is located.
"""
import argparse
import collections
import os
import sys
import unittest


_VERBOSITY_LOWER_LIMIT = 0
_VERBOSITY_UPPER_LIMIT = 3
_TOP_LEVEL_TEST_MODULE = "rwlock"

CmdLineArgs = collections.namedtuple("CmdLineArgs",
                                     ["verbosity"])  # int


def _add_code_root():
    """
    Adds the root of the code tree (that is being tested) to the head of sys.path.

    Note that it is important to add the *code* tree root ahead of the *test* tree root
    (which will be added automatically since this script is intended to be invoked from
    the test tree root), because we might have a hierarchy within the test tree that looks
    identical, in terms of package/module names, to the code tree.  Hence, in order for
    imports within the test code to be able to load the proper code that is being tested
    (and not try to find corresponding names/modules/classes/etc. from the *test* tree,
    which are obviously not there), the path to the code tree must be ahead in the search path.
    """
    parent_dir_of_this_script = os.path.dirname(os.path.realpath(__file__))
    path_to_main_code_root = os.path.realpath(os.path.join(parent_dir_of_this_script, ".."))
    sys.path.insert(0, path_to_main_code_root)


def _parse_command_line():
    """
    Parses the command line for this script.

    @return CmdLineArgs: An object encapsulating the values for the command line args.
    """
    # Customizing the help formatter used with a max_help_position larger than the current
    # default of 24 gives a little more room in the help printout, between the parameter/value
    # list and the help string:
    def formatter(prog): return argparse.ArgumentDefaultsHelpFormatter(prog,
                                                                       max_help_position=36)

    # noinspection PyTypeChecker
    parser = argparse.ArgumentParser(
        formatter_class=formatter,  # The above-defined formatter
        description="Runs unit tests for rwlock.")
    parser.add_argument("-v", "--verbosity",
                        metavar="VAL",
                        type=int,
                        default="2",
                        help="the value for test output verbosity; must be a non-negative "
                             "integer in between {0} and {1}".format(_VERBOSITY_LOWER_LIMIT,
                                                                     _VERBOSITY_UPPER_LIMIT))
    args = parser.parse_args()
    return CmdLineArgs(verbosity=args.verbosity)


def _run_tests(verbosity):
    """
    Creates the test-suite through test discovery from the root of the test tree, and runs it.

    NOTE:
        - Test suite dirs must look like python modules, i.e., all file names must be valid
          Python identifiers (e.g., there shouldn't be any hyphens in the file names), and
          the dir must have a __init__.py file.
        - Test class files must end in the suffix "_test.py".
        - Test method names in test classes must start with the prefix "test_".

    @param int verbosity: The verbosity of the test output.
    @return unittest.result.TestResult: The test result.
    """
    loader = unittest.TestLoader()
    tests = loader.discover(_TOP_LEVEL_TEST_MODULE, pattern="*_test.py")
    # noinspection PyTypeChecker
    test_runner = unittest.TextTestRunner(stream=sys.stdout, verbosity=verbosity)
    # noinspection PyUnresolvedReferences
    return test_runner.run(tests)


def main():
    """The main entry point."""
    _add_code_root()
    cmd_line_args = _parse_command_line()
    # The following import only works after _add_code_root():
    print("Running tests for rwlock ...")
    print("[Using {}]".format(sys.executable))
    test_result = _run_tests(cmd_line_args.verbosity)
    if test_result.wasSuccessful():
        return 0
    else:
        return 1


if __name__ == "__main__":
    sys.exit(main())


test/rwlock/rwlock_test.py
# Copyright (C) 2020 Ankan Pramanick - All rights reserved.
import threading
import time
import unittest
import sys
import io
import re
import random

from rwlock import SHLock

_TYPE_READER = 0
_TYPE_WRITER = 1
_WRITER_OUTPUT_PATTERN = re.compile(r"^\[Writer")


class _TestThread1(threading.Thread):
    """A thread class to use in our unit tests."""

    _SHARED_RESOURCE = "0"
    _SHARED_RESOURCE_LOCK = SHLock()

    def __init__(self, thread_type, thread_id, name, total_step_count, step_delay):
        """Ctor.

        @param int thread_type:      Thread type, one of _TYPE_READER or _TYPE_WRITER.
        @param int thread_id:        Thread ID.
        @param str name:             Thread name.
        @param int total_step_count: Number of times the run() method should do some work.
        @param float step_delay:     The delay in seconds between the work steps in the run()
                                     method.
        """
        threading.Thread.__init__(self, name=name)
        self.type = thread_type
        self.id = thread_id
        self.name = name
        self.counter = total_step_count
        self.step_delay = step_delay

    def run(self):
        append_char = None
        if _TYPE_WRITER == self.type:
            append_char = "{}".format(self.id)
        print("Starting {}".format(self.name))
        while self.counter > 0:
            time.sleep(self.step_delay)
            self.counter -= 1
            got_lock = False
            if _TYPE_READER == self.type:
                try:
                    _TestThread1._SHARED_RESOURCE_LOCK.acquire(shared=True)
                    got_lock = True
                    print("[{}] Resource value: [{}]".format(self.name,
                                                             _TestThread1._SHARED_RESOURCE))
                except Exception as e:
                    self.counter = 0
                    print("[{}] [error] Aborting because of exception: {}".format(self.name, e))
                finally:
                    if got_lock:
                        _TestThread1._SHARED_RESOURCE_LOCK.release()
            else:  # Writer
                try:
                    _TestThread1._SHARED_RESOURCE_LOCK.acquire(shared=False)
                    got_lock = True
                    print("[{}] Updating resource with '{}' ...".format(self.name, append_char))
                    _TestThread1._SHARED_RESOURCE = _TestThread1._SHARED_RESOURCE + append_char
                except Exception as e:
                    self.counter = 0
                    print("[{}] [error] Aborting because of exception: {}".format(self.name, e))
                finally:
                    if got_lock:
                        _TestThread1._SHARED_RESOURCE_LOCK.release()
        print("Exiting {}".format(self.name))


class _TestThread2(threading.Thread):
    """A thread class to use in our unit tests."""

    _SHARED_RESOURCE_LOCK = SHLock()
    _WRITER_SLEEP_TIME = 0.5  # seconds

    def __init__(self, thread_type, thread_id, name, outfile=sys.stdout):
        """Ctor.

        @param int thread_type: Thread type, one of _TYPE_READER or _TYPE_WRITER.
        @param int thread_id:   Thread ID.
        @param str name:        Thread name.
        @param stream outfile:  Object to stream output to.
        """
        threading.Thread.__init__(self, name=name)
        self.type = thread_type
        self.id = thread_id
        self.name = name
        self.outfile = outfile

    def run(self):
        # print("{} starting".format(self.name))
        got_lock = False
        if _TYPE_READER == self.type:
            try:
                _TestThread2._SHARED_RESOURCE_LOCK.acquire(shared=True)
                got_lock = True
                print("[{}] Acquired lock".format(self.name), file=self.outfile)
                # Pause enough for other readers a chance to get the shared lock while we are
                # still holding on to it:
                delay = (random.randint(1, 4)) / 100.0
                time.sleep(delay)
            except Exception as e:
                print("[{}] [error] Aborting because of exception: {}".format(self.name, e))
            finally:
                if got_lock:
                    _TestThread2._SHARED_RESOURCE_LOCK.release()
                    print("[{}] Released lock".format(self.name), file=self.outfile)
        else:  # Writer
            try:
                _TestThread2._SHARED_RESOURCE_LOCK.acquire(shared=False)
                got_lock = True
                print("[{}] Acquired lock".format(self.name), file=self.outfile)
                time.sleep(_TestThread2._WRITER_SLEEP_TIME)
            except Exception as e:
                print("[{}] [error] Aborting because of exception: {}".format(self.name, e))
            finally:
                if got_lock:
                    _TestThread2._SHARED_RESOURCE_LOCK.release()
                    print("[{}] Released lock".format(self.name), file=self.outfile)
        # print("{} exiting".format(self.name))


class RWLockTest(unittest.TestCase):
    """Unit tests for the rwlock module.

    Note that currently we don't really have comprehensive unit tests for rwlock;
    we just have a few simple tests that trace the execution of several competing
    threads when SHLock is used in both modes.
    """

    def test_simple(self):
        r1 = _TestThread1(_TYPE_READER, 1, "Reader 1", 25, 0.1)
        r2 = _TestThread1(_TYPE_READER, 2, "Reader 2", 15, 0.15)
        r3 = _TestThread1(_TYPE_READER, 3, "Reader 3", 4, 0.5)
        w1 = _TestThread1(_TYPE_WRITER, 4, "Writer 1", 5, 0.05)
        w2 = _TestThread1(_TYPE_WRITER, 5, "Writer 2", 7, 0.3)
        all_threads = [r1, r2, r3, w1, w2]
        for t in all_threads:
            t.start()
        for t in all_threads:
            t.join(timeout=5)
            self.assertFalse(t.is_alive())

    def test_exclusive_lockout_writer_first(self):
        with io.StringIO() as output:
            w0 = _TestThread2(_TYPE_WRITER, 0, "Writer 0", output)
            r1 = _TestThread2(_TYPE_READER, 1, "Reader 1", output)
            r2 = _TestThread2(_TYPE_READER, 2, "Reader 2", output)
            r3 = _TestThread2(_TYPE_READER, 3, "Reader 3", output)
            r4 = _TestThread2(_TYPE_READER, 4, "Reader 4", output)
            r5 = _TestThread2(_TYPE_READER, 5, "Reader 5", output)
            all_threads = [w0, r1, r2, r3, r4, r5]
            w0.start()  # Start the writer first: it should be able to grab the lock exclusively
            time.sleep(0.01)
            # None of reader threads should be able to acquire the lock until after
            # the writer releases it:
            for t in all_threads:
                if t is not w0:
                    t.start()
            for t in all_threads:
                t.join(timeout=2)
                self.assertFalse(t.is_alive())
            # The first two lines in the captured output should always be from the writer:
            lines = output.getvalue().splitlines()
            self.assertEqual(12, len(lines))
            line0 = lines[0].strip()
            line1 = lines[1].strip()
            self.assertEqual("[Writer 0] Acquired lock", line0)
            self.assertEqual("[Writer 0] Released lock", line1)

    def test_exclusive_lockout_writer_later(self):
        with io.StringIO() as output:
            w0 = _TestThread2(_TYPE_WRITER, 0, "Writer 0", output)
            r1 = _TestThread2(_TYPE_READER, 1, "Reader 1", output)
            r2 = _TestThread2(_TYPE_READER, 2, "Reader 2", output)
            r3 = _TestThread2(_TYPE_READER, 3, "Reader 3", output)
            r4 = _TestThread2(_TYPE_READER, 4, "Reader 4", output)
            r5 = _TestThread2(_TYPE_READER, 5, "Reader 5", output)
            all_threads = [r1, r2, w0, r3, r4, r5]
            for t in all_threads:
                time.sleep(0.001)
                t.start()
            for t in all_threads:
                t.join(timeout=2)
                self.assertFalse(t.is_alive())
            lines = output.getvalue().splitlines()
            self.assertEqual(12, len(lines))
            # There should be no "[Reader ..." lines in between the two "[Writer ..." lines:
            for i in range(len(lines)):
                line = lines[i].strip()
                if _WRITER_OUTPUT_PATTERN.search(line):
                    self.assertLess(i, len(lines) - 1)
                    next_line = lines[i + 1]
                    self.assertIsNotNone(_WRITER_OUTPUT_PATTERN.search(next_line))
                    break


rwlock/__init__.py
"""

rwlock: A Python implementation of a readers/writer lock.

This is a drop-in replacement for the threading2.SHLock class, using the implementation
from that library.  The threading2 library (https://github.com/rfk/threading2) is no longer
maintained (its github repository has been archived), and is stuck at Python 2.x.  This
implementation has been ported to work with Python 3.8.

Note that the SHLock class is an implementation of the classic readers-writer lock (see,
for example, https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock).  A read/write lock
allows concurrent access for read-only operations, while write operations require exclusive
access.  This means that multiple threads can read the data in parallel, but an exclusive
lock is needed for writing or modifying data. When a writer is writing the data, all other
writers or readers will be blocked until the writer is finished writing.

The following API niceties are also included:

    * all blocking methods take a "timeout" argument and return a success code
    * all exposed objects are actual classes and can be safely subclassed

"""
import sys
_MIN_PYTHON = (3, 8)
if sys.version_info < _MIN_PYTHON:
    sys.exit("Python {}.{} or later is required.".format(_MIN_PYTHON[0], _MIN_PYTHON[1]))

__ver_major__ = 0
__ver_minor__ = 5
__ver_patch__ = 1
__ver_sub__ = ""
__version__ = "{0}.{1}.{2}{3}".format(__ver_major__, __ver_minor__, __ver_patch__, __ver_sub__)

from rwlock.rw_lock import *

__all__ = ["Condition", "Lock", "RLock", "SHLock"]


rwlock/rw_lock.py
import threading
import time

__all__ = ["Condition", "Lock", "RLock", "SHLock"]


# noinspection PyUnresolvedReferences
class _ContextManagerMixin(object):
    """Simple mixin mapping __enter__/__exit__ to acquire/release."""

    def __enter__(self):
        self.acquire()
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.release()


class Lock(_ContextManagerMixin):
    """Class-based Lock object.

    This is a very thin wrapper around Python's native lock objects.  It's
    here to provide ease of subclassing and to add a "timeout" argument
    to Lock.acquire().
    """

    # noinspection PyUnresolvedReferences,PyProtectedMember
    def __init__(self):
        self._lock = threading._allocate_lock()
        super(Lock, self).__init__()

    def acquire(self, blocking=True, timeout=None):
        """Attempt to acquire this lock.

        If the optional argument "blocking" is True and "timeout" is None,
        this methods blocks until is successfully acquires the lock.  If
        "blocking" is False, it returns immediately if the lock could not
        be acquired.  Otherwise, it blocks for at most "timeout" seconds
        trying to acquire the lock.

        In all cases, this methods returns True if the lock was successfully
        acquired and False otherwise.
        """
        if timeout is None:
            return self._lock.acquire(blocking)
        else:
            # Simulated timeout using progressively longer sleeps.
            # This is the same timeout scheme used in the stdlib Condition
            # class.  If there's lots of contention on the lock then there's
            # a good chance you won't get it; but then again, Python doesn't
            # guarantee fairness anyway.
            end_time = time.time() + timeout
            delay = 0.0005
            while not self._lock.acquire(False):
                remaining = end_time - time.time()
                if remaining <= 0:
                    return False
                delay = min(delay * 2, remaining, 0.05)
                time.sleep(delay)
            return True

    def release(self):
        """Release this lock."""
        self._lock.release()


# noinspection PyProtectedMember
class RLock(_ContextManagerMixin, threading._RLock):
    """Re-implemented RLock object.

    This is pretty much a direct clone of the RLock object from the standard
    threading module; the only difference is that it uses a custom Lock class
    so that acquire() has a "timeout" parameter.

    NOTE(ankan): So why the inheritance from _RLock at all?  That is because Condition
    (see below) uses *this* RLock, while inheriting from threading.Condition, and the
    latter needs the *other* methods in _RLock like _acquire_restore() and _release_save(),
    which are not overridden here.
    """

    _LockClass = Lock

    def __init__(self):
        super(RLock, self).__init__()
        self._block = self._LockClass()
        self._owner = None
        self._count = 0

    def acquire(self, blocking=True, timeout=None):
        me = threading.get_ident()
        if self._owner == me:
            self._count += 1
            return True
        if self._block.acquire(blocking, timeout):
            self._owner = me
            self._count = 1
            return True
        return False

    def release(self):
        if self._owner != threading.get_ident():
            raise RuntimeError("cannot release un-acquired lock")
        self._count -= 1
        if not self._count:
            self._owner = None
            self._block.release()

    def _is_owned(self):
        return self._owner == threading.get_ident()


class Condition(threading.Condition):
    """Re-implemented Condition class.

    This is pretty much a direct clone of the Condition class from the standard
    threading module; the only difference is that it uses a custom Lock class
    so that acquire() has a "timeout" parameter.
    """

    _LockClass = RLock
    _WaiterLockClass = Lock

    def __init__(self, lock=None):
        if lock is None:
            lock = self._LockClass()
        super(Condition, self).__init__(lock)

    # This is essentially the same as the base version, but it returns
    # True if the wait was successful and False if it timed out.
    #
    # NOTE(ankan): This is not strictly necessary in Py 3.8+, since there,
    # threading.Condition.wait() already returns the above values under the same
    # conditions.  However, we continue to use this re-implementation, since we
    # are using our custom classes for _LockClass and _WaiterLockClass.
    # noinspection PyUnresolvedReferences
    def wait(self, timeout=None):
        if not self._is_owned():
            raise RuntimeError("cannot wait on un-acquired lock")
        waiter = self._WaiterLockClass()
        waiter.acquire()
        self._waiters.append(waiter)
        saved_state = self._release_save()
        try:
            if not waiter.acquire(timeout=timeout):
                try:
                    self._waiters.remove(waiter)
                except ValueError:
                    pass
                return False
            else:
                return True
        finally:
            self._acquire_restore(saved_state)


class SHLock(_ContextManagerMixin):
    """Shareable lock class.

    This functions just like an RLock except that you can also request a
    "shared" lock mode.  Shared locks can co-exist with other shared locks
    but block exclusive locks.  You might also know this as a read/write lock.

    Currently attempting to upgrade or downgrade between shared and exclusive
    locks will cause a deadlock.  This restriction may go away in future.
    """

    class Context(_ContextManagerMixin):

        def __init__(self, parent,
                     blocking=True, timeout=None, shared=False):
            self.parent = parent
            self.blocking = blocking
            self.timeout = timeout
            self.shared = shared

        def acquire(self):
            self.parent.acquire(blocking=self.blocking,
                                timeout=self.timeout,
                                shared=self.shared)

        def release(self):
            self.parent.release()

    _LockClass = Lock
    _ConditionClass = Condition

    def __init__(self):
        self._lock = self._LockClass()
        # When a shared lock is held, is_shared will give the cumulative
        # number of locks and _shared_owners maps each owning thread to
        # the number of locks is holds.
        self.is_shared = 0
        self._shared_owners = {}
        # When an exclusive lock is held, is_exclusive will give the number
        # of locks held and _exclusive_owner will give the owning thread
        self.is_exclusive = 0
        self._exclusive_owner = None
        # When someone is forced to wait for a lock, they add themselves
        # to one of these queues along with a "waiter" condition that
        # is used to wake them up.
        self._shared_queue = []
        self._exclusive_queue = []
        # This is for recycling waiter objects.
        self._free_waiters = []

    def __call__(self, blocking=True, timeout=None, shared=False):
        return SHLock.Context(self, blocking=blocking,
                              timeout=timeout, shared=shared)

    def acquire(self, blocking=True, timeout=None, shared=False):
        """Acquire the lock in shared or exclusive mode."""
        with self._lock:
            if shared:
                self._acquire_shared(blocking, timeout)
            else:
                self._acquire_exclusive(blocking, timeout)
            assert not (self.is_shared and self.is_exclusive)

    def release(self):
        """Release the lock."""
        # This decrements the appropriate lock counters, and if the lock
        # becomes free, it looks for a queued thread to hand it off to.
        # By doing the hand-off here we ensure fairness.
        me = threading.currentThread()
        with self._lock:
            if self.is_exclusive:
                if self._exclusive_owner is not me:
                    raise RuntimeError("release() called on un-acquired lock")
                self.is_exclusive -= 1
                if not self.is_exclusive:
                    self._exclusive_owner = None
                    # If there are waiting shared locks, issue it to them
                    # all and then wake everyone up.
                    if self._shared_queue:
                        for (thread, waiter) in self._shared_queue:
                            self.is_shared += 1
                            self._shared_owners[thread] = 1
                            waiter.notify()
                        del self._shared_queue[:]
                    # Otherwise, if there are waiting exclusive locks,
                    # they get first dibs on the lock.
                    elif self._exclusive_queue:
                        (thread, waiter) = self._exclusive_queue.pop(0)
                        self._exclusive_owner = thread
                        self.is_exclusive += 1
                        waiter.notify()
            elif self.is_shared:
                try:
                    self._shared_owners[me] -= 1
                    if self._shared_owners[me] == 0:
                        del self._shared_owners[me]
                except KeyError:
                    raise RuntimeError("release() called on un-acquired lock")
                self.is_shared -= 1
                if not self.is_shared:
                    # If there are waiting exclusive locks, they get first dibs on the lock.
                    if self._exclusive_queue:
                        (thread, waiter) = self._exclusive_queue.pop(0)
                        self._exclusive_owner = thread
                        self.is_exclusive += 1
                        waiter.notify()
                    else:
                        assert not self._shared_queue
            else:
                raise RuntimeError("release() called on un-acquired lock")

    def _acquire_shared(self, blocking=True, timeout=None):
        me = threading.currentThread()
        # Each case: acquiring a lock we already hold.
        if self.is_shared and me in self._shared_owners:
            self.is_shared += 1
            self._shared_owners[me] += 1
            return True
        # If the lock is already spoken for by an exclusive, add us
        # to the shared queue and it will give us the lock eventually.
        if self.is_exclusive or self._exclusive_queue:
            if self._exclusive_owner is me:
                raise RuntimeError("can't downgrade SHLock object")
            if not blocking:
                return False
            waiter = self._take_waiter()
            try:
                self._shared_queue.append((me, waiter))
                if not waiter.wait(timeout=timeout):
                    self._shared_queue.remove((me, waiter))
                    return False
                assert not self.is_exclusive
            finally:
                self._return_waiter(waiter)
        else:
            self.is_shared += 1
            self._shared_owners[me] = 1

    def _acquire_exclusive(self, blocking=True, timeout=None):
        me = threading.currentThread()
        # Each case: acquiring a lock we already hold.
        if self._exclusive_owner is me:
            assert self.is_exclusive
            self.is_exclusive += 1
            return True
        # If the lock is already spoken for, add us to the exclusive queue.
        # This will eventually give us the lock when it's our turn.
        if self.is_shared or self.is_exclusive:
            if not blocking:
                return False
            waiter = self._take_waiter()
            try:
                self._exclusive_queue.append((me, waiter))
                if not waiter.wait(timeout=timeout):
                    self._exclusive_queue.remove((me, waiter))
                    return False
            finally:
                self._return_waiter(waiter)
        else:
            self._exclusive_owner = me
            self.is_exclusive += 1

    def _take_waiter(self):
        try:
            return self._free_waiters.pop()
        except IndexError:
            return self._ConditionClass(self._lock)

    def _return_waiter(self, waiter):
        self._free_waiters.append(waiter)
